src/modules/keyboardmanager/common/KeyboardManagerState.cpp

```diff
@@ -181,14 +183,35 @@ void KeyboardManagerState::UpdateDetectShortcutUI()
     detectedShortcut_lock.unlock();
 
     // Since this function is invoked from the back-end thread, in order to update the UI the dispatcher must be used.
-    currentShortcutUI.Dispatcher().RunAsync(Windows::UI::Core::CoreDispatcherPriority::Normal, [this, detectedShortcutCopy]() {
+    currentShortcutUI1.Dispatcher().RunAsync(Windows::UI::Core::CoreDispatcherPriority::Normal, [this, detectedShortcutCopy]() {
         std::vector<hstring> shortcut = detectedShortcutCopy.GetKeyVector(keyboardMap);
-        currentShortcutUI.Children().Clear();
+        currentShortcutUI1.Children().Clear();
+        currentShortcutUI2.Children().Clear();
+
+        // The second row should be hidden if there are 3 keys or lesser to avoid an extra margin
+        if (shortcut.size() > 3)
+        {
+            currentShortcutUI2.Visibility(Visibility::Visible);
+        }
+        else
+        {
+            currentShortcutUI2.Visibility(Visibility::Collapsed);
+        }
+
+        int i = 0;
         for (auto& key : shortcut)
         {
-            AddKeyToLayout(currentShortcutUI, key);
+            if (i < 3)
+            {
+                AddKeyToLayout(currentShortcutUI1, key);
+            }
+            else
+            {
+                AddKeyToLayout(currentShortcutUI2, key);
+            }
+            i++;
         }
-        currentShortcutUI.UpdateLayout();
+        currentShortcutUI1.UpdateLayout();
     });

```

Comment:
Maybe change this to `for(int i=0; i < shortcut.size(); i++)` 
Also shouldn't  `currentShortcutUI2.UpdateLayout()` be called as well?

Reply:
Good catch. I'll make that change. I've notice sometimes an explicit updatelayout is not always required which is maybe why it was working without it, but I'll add it to be on the safe side.